### ISSUE TITLE
Proposal: ETL language

### ISSUE THREAD

**Issue by sbelak** [metabase/metabase#9847]
# ETL language

## Language spec

The language has 3 sections: `requires`, `provides`, and `steps`.

### requires

Defines and names all the tables (and/or _data models_) needed. The pipeline applies only if all are present in full (ie all the dimensions). Binding is done based on two criteria: fields present (`dimensions`) and/or entity type (`entity_type`). We might also want to support constraints such as `links_to`/`linked_from`

### provides

Lists resulting data models. 

### steps

Each step will result in a card. The entire pipeline will be in its own (sub)collection in `ETL` collection.

## Example:
```
name: MRR & Churn
description: Builds tables needed for MRR and churn calculations
requires:
  OrderHistory:
      dimensions:
        - Income
        - CreatedAt
	- User
provides:
  MRRStatistics:
       dimensions:
         - Month
         - ChurrnedMRR
         - MRRChurn
         - AccountChurn
         - MRR
         - Accounts
         - ARPU
steps:
  Charges:
      source: OrderHistory
      expressions:
        Month: [date-time, [dimension, CreatedAt], month]
	PrevMonth: [-, [dimension, Month], [interval, -1, month]]
  NextMonth
       source: Charges
  MRR:
      source: Charges
      aggregation:
        MRR: [sum, [dimension, Income]]
	Accounts: [count-distinct, [dimension, User]]
	ARPU: [/, [aggregation, MRR], [aggregation, Accounts]]
      breakout: Month
  MRRStatistics:
      source: Charges
      expressions:
        Month: [-, [dimension, Month], [interval, 1, month]]
      aggregation:
        ChurrnedMRR: [sum-when, [is-null, [dimension, NextMonth.Month]], [dimension, Total]]
	MRRChurn: [/, [dimension, ChurrnedMRR], [dimension, MRR.MRR]]
	AccountChurn: [count-when, [is-null, [dimension, NextMonth.Month]]]
      breakout:
        - Month
	- MRR.MRR
	- MRR.Accounts
      join:
        - [left-join, Charges, [and, [=, [dimension, Charges.User], [dimension, NextMonth.User]], [=, [dimension, Charges.Month], [dimension, NextMonth.PrevMonth]]]]
        - [inner-join, MRR, [dimension, Charges.Month], [dimension, MRR.Month]]
```

## Open questions, annoyances & notes
* Writing MBQL in YAML is annoying
* Wrapping entities in `dimension`/`aggregation` is annoying, but it massively simplifies generation of the actual MBQL.
* The language (similarly to x-ray heuristics) language somewhat conflates names and types. The thinking behind is that our types are "true names", but we can easily add an explicit binding step like we do with x-rays.
* What to do if multiple tables match a single requires binding? Should we consider no matches then, do a cartesian product of all possible matches, have control over the `exactly one` vs `every one` semantics? Ditto for dimensions. 
* Ideally this would leverage MBQL joins, but for v1 we can also spit out raw SQL.

**Comment by tlrobinson** [comment]
This looks a lot like MBQL with a few differences:

1. slightly different names (`aggregations` instead of `aggregation`), etc
2. `[dimension ...]` instead of something like `[field-literal ...]`
3. maps of names to aggregations/breakouts instead of `["named", aggregation, "Name"]` etc

Maybe we just want MTBL to be a special human-readable/writable YAML serialization of MBQL? It doesn’t have to a simple 1-to-1 YAML serialization of MBQL (that would be incredibly verbose), it could incorporate some of these niceties, but maybe the underlying model should be MBQL?

**Comment by sbelak** [comment]
1. was just me being sloppy (although our current inconsistent pluralization is jarring)
2. dimension is a thing you bind/match, not directly a reference to some entity in the DB, hence a different name (also to make things easier when parsing, but that's secondary)
3. is done 3 reasons: consistency (:expressions is already a map in MBQL), it's nicer to write in YAML, and it's nicer to parse 

The idea is that MBTL gets compiled into a series of MBQL cards so the correspondence between the two will always be strong, but MBTL can have some niceties that take advantage of scope being more limited/us having more context information and also just to make writing it by hand nicer.

**Comment by sbelak** [comment]
Alternative join syntax:
```
join:
        - Charges: [left-join, [and, [=, [dimension, Charges.User], [dimension, NextMonth.User]], [=, [dimension, Charges.Month], [dimension, NextMonth.PrevMonth]]]]
	- MRR: [inner-join, [dimension, Charges.Month], [dimension, MRR.Month]]
```
Have a clause for each join type with 2 airities: the 1-airity takes a filter clause, the 2-airity takes 2 dimensions and joins them on =.

**Comment by sbelak** [comment]
Updated with `provides` and the new join syntax

**Comment by sbelak** [comment]
Should be constraint resulting models to be comprised only of known types? It would be kind of neat, but it also means we have to have types for very specific things such as pre-aggregations. Also forcing  `Month` to be a type is somewhat awkward. and to a certain degree it duplicates the `:unti` functionality, although not completely as it represents the maximal resolution and can be the result of truncation, while `:unit` stands for bucketing (and implies aggregation).

**Comment by Luckymethod** [comment]
I'm not in agreement that the investment in building anything in this area would be valuable considering how much ground Metabase needs to make in regarding to bread and butter BI and visualization features. Also, why creating a new YAML-based language (yuck) when what you want to get done is easier and more expressive to get done in python using Airflow? Or any other tool of that nature. If I'm missing something please help me understand.

**Comment by sbelak** [comment]
@Luckymethod: this will not necessarily be exposed to the end users, but is needed to allow more advanced automatic analysis such as x-rays where we're currently too constrained with what shape the user's data is. As a bonus, implementing transform steps as Metabase questions gives you fantastic observability and sidesteps all the infrastructure blues you otherwise have with ETL.

MBTL is just a thin DSL over our existing MBQL and as such purely code-is-data (I share your distaste for YAML so in my mind this is EDN/Clojure). The reason we need MBQL is to be SQL-dialect agnostic (this is a constraint most people building ETL systems don't have). We also very much don't want to bring in new infrastructure requirements such as Airflow or having a Py runtime.

**Comment by sbelak** [comment]
```
name: Subscriptions
description: Subscription log
refines: Generic Dataset
required_attributes:
  - field: StartTime
  - field: EndTime
  - data_model: User
  - has_many:
      data_model: RenewalEvent
optional_attributes:
  - data_model: Plan
  - field: Price
metrics:
   ActiveSubscriptions:
     aggregation: [count]
     description: ...
     breakouts:
       - Plan
       - NewUsers
       - Paying
   MRR:
     aggregation: [sum, [dimension, Price]]
     filter: [>, [dimension, EndTime], [relative-datetime current] ]
     description: Monthly Recurring Revenue
segments:
  NewUsers:
    filter: [time-interval, [dimension, StartTime], -30, days]
    description: ...
  Paying:
     filter: [>, [dimension, Price], 0]
     description: ...
  YearlyPlans:
     filter: [>=, [-, [dimension, EndTime], [dimension, StartTime]] 12]
     description: ...
breakout_dimensions:
  - Plan.name
  - StartTime
```
Generalized *_dimensions to *_attributes which allow modelling relations as well. 

In case of one-to-one mapping the binding logic should be smart enough to also work with a denomralized table (eg. Subscriptions and User parts are co-located in the same materialized table)

**Comment by sbelak** [comment]
Implemented in #10105
